Upgrade to Clang 5.0.0

amaitland opened the pull request:
I've run the Generator against `Clang 5.0.0` and this is the result

- Removed `index_setClientEntity` and `index_getClientEntity` from generated code, same as https://github.com/Microsoft/ClangSharp/pull/16. If this is incorrect then pointers welcome.
- Added additional check for `CXTypeKind.CXType_Elaborated` as per https://github.com/Microsoft/ClangSharp/pull/16 this can be easily reverted if required.

The `Generated.cs` and `Generated.Custom.cs` files contained within `ClangSharpPInvokeGenerator` are different to those in root, is there a reason they're different? I've updated them so they're now the same, this can easily be reverted.

I'm interested in a `Clang 5.0.0` release and I'm happy to contribute where possible. I've kept the three distinct commits as it's now very easy to `Squash Merge` using the `GitHub` interface and I personally feel it makes the changes easy to review. Feedback welcome.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/ClangSharp?pullRequest=18) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/ClangSharp?pullRequest=18) before we can accept your contribution.<br/><br/>:x: amaitland [sign now](https://cla.opensource.microsoft.com/Microsoft/ClangSharp?pullRequest=18)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/ClangSharp?pullRequest=18) it.</sub>

amaitland commented:
Happy to sign the `cla` when I'm next back in front of a `PC`.

amaitland commented:
> Removed index_setClientEntity and index_getClientEntity from generated code, same as #16. If this is incorrect then pointers welcome.

Looked into this further and appears the generated ones should be excluded, along with a number of other generated functions, these are just the ones that clashed. Added `--excludeFunctions` command line argument and updated the batch file to exclude what are hopefully all the custom generated functions.

> Added additional check for CXTypeKind.CXType_Elaborated as per #16 this can be easily reverted if required.

Debugging this a little more, the output appears to be the same, I think it's a minor improvement when debugging, so keeping it seems reasonable.

I've added `--clangPath` command line argument to make life easier when debugging, adding the ability to load `libclang` from a different directory.

amaitland commented:
> Debugging this a little more, the output appears to be the same, I think it's a minor improvement when debugging, so keeping it seems reasonable.

Changed my mind, removed as it's unnecessary

mjsabby commented on ClangSharpPInvokeGenerator/FunctionVisitor.cs:
```diff
@@ -13,12 +13,21 @@ internal sealed class FunctionVisitor : ICXCursorVisitor
         private readonly string prefixStrip;
 
-        public FunctionVisitor(TextWriter tw, string libraryPath, string prefixStrip)
+        public FunctionVisitor(TextWriter tw, string libraryPath, string prefixStrip, string[] execludeFunctionsArray)
```
spelling

amaitland replied:
Fixed, thx

mjsabby commented on ClangSharpPInvokeGenerator/FunctionVisitor.cs:
```diff
@@ -13,12 +13,21 @@ internal sealed class FunctionVisitor : ICXCursorVisitor
+
+            if(execludeFunctionsArray != null)
+            {
+                //Fir all the functions we're excluding add them to the visitedFunctions set so they're ignored
```
spelling: Fire.

amaitland replied:
Fixed, thx

mjsabby commented on ClangSharpPInvokeGenerator/Generated.Custom.cs:
```diff
@@ -117,30 +117,38 @@ public static IntPtr codeCompleteAt(CXTranslationUnit @TU, string @complete_file
             {
                 BeginCXUnsavedFileMarshal(ref arr, ref unsaved_files);
-                return indexSourceFile(param0, client_data, index_callbacks, index_callbacks_size, index_options, source_filename, command_line_args, num_command_line_args, arr, num_unsaved_files, out out_TU, TU_options);
+                BeginIndexerCallbacksMarshal(out result, ref index_callbacks);
```
I'm assuming this was a bug in the original code and you're having to marshal this correctly?

amaitland replied:
This whole commit is simply the result of copying `Generated.cs` and `Generated.Custom.cs` from the root over the top of  `ClangSharpPInvokeGenerator\Generated.cs` and `ClangSharpPInvokeGenerator\Generated.Custom.cs`.

Just making them consistent. Happy to revert this whole commit if you'd prefer?

amaitland replied:
Thinking about this more, this is probably my mistake, I probably synced the files in the incorrect direction....

amaitland replied:
Maybe the two of the file should be removed from `Git` and copied as part of the build/post build process? (Assuming `VS` plays nice with the missing files)

Unless I'm missing something and there's a reason they're different.

mjsabby commented on ClangSharpPInvokeGenerator/Generated.Custom.cs:
```diff
@@ -117,30 +117,38 @@ public static IntPtr codeCompleteAt(CXTranslationUnit @TU, string @complete_file
                 BeginCXUnsavedFileMarshal(ref arr, ref unsaved_files);
-                return indexSourceFileFullArgv(param0, client_data, index_callbacks, index_callbacks_size, index_options, source_filename, command_line_args, num_command_line_args, arr, num_unsaved_files, out out_TU, TU_options);
+                BeginIndexerCallbacksMarshal(out result, ref index_callbacks);
+
```
extra linebreak.

mjsabby commented on ClangSharpPInvokeGenerator/Program.cs:
```diff
@@ -85,11 +107,24 @@ public static void Main(string[] args)
+                Environment.Exit(-1);
+            }
+
+            if(!string.IsNullOrEmpty(libClangPath) && File.Exists(libClangPath) && (libClangPath.EndsWith(".dll") || libClangPath.EndsWith(".so")))
```
why do you need this?

amaitland replied:
Which part? Would you prefer a console message is written if `libclang` doesn't exist? Or are you referring to the feature in general? 

I admit the code it's self is a little on the defensive side, makes sure that the file exists, that it's a `dll` or `so`, all seem reasonable checks, though looking back maybe writing a message to the console and terminating execution would be better?

mjsabby commented on ClangSharpPInvokeGenerator/Generated.cs:
```diff
@@ -1911,13 +1911,13 @@ public static partial class clang
 
         [DllImport(libraryPath, EntryPoint = "clang_annotateTokens", CallingConvention = CallingConvention.Cdecl)]
-        public static extern void annotateTokens(CXTranslationUnit @TU, CXToken[] @Tokens, uint @NumTokens, out CXCursor @Cursors);
+        public static extern void annotateTokens(CXTranslationUnit @TU, out CXToken @Tokens, uint @NumTokens, out CXCursor @Cursors);
```
any reason why you've removed the [] and made it out?

amaitland replied:
Direct copy of  https://github.com/Microsoft/ClangSharp/blob/master/Generated.cs#L1917

amaitland commented:
Removed the unnecessary commit that synced the `Generated.cs` and `Generated.Custom.cs` files, was incorrect.

benpye commented:
@mjsabby Is there any reason this hasn't been looked at again? Looking to use ClangSharp but binding to such an old version as 3.x is a blocker. Even this is somewhat out of date now with 6.0.0 released.

amaitland commented:
Closing in favor of #21
